### ISSUE TITLE
Odd bug with character handling

### ISSUE THREAD

**Issue by pmoura** [infradig/trealla#227]
There's an odd bug where instead of an expected char we get the empty atom. We see it on the `write_term/3` tests:

```text
!     lgt_write_term_3_25: failure (in 0.0015849999999999476 seconds)
!       test assertion failed: 'A'=='\'\''
!       in file /Users/pmoura/logtalk/tests/prolog/predicates/write_term_3/tests.lgt between lines 142-148
```

But if we comment out all the other tests, the test passes.

On the `base64` library tests:

```text
!     base64_generate_2_01: failure (in 0.0010470000000000201 seconds)
!       test assertion failed: 'VhlIF1aWNrIJyb3duIZveqdW1wcyvdmVyIRoZSsYXp5IRvZy4='=='VGhlIHF1aWNrIGJyb3duIGZveCBqdW1wcyBvdmVyIHRoZSBsYXp5IGRvZy4='
!       in file /Users/pmoura/logtalk/library/base64/tests.lgt between lines 51-56
!     base64_generate_2_02: failure (in 0.0014279999999997628 seconds)
!       test assertion failed: ['V','',h,l,'I','','F','1',a,'W','N',r,'I','','J',y,b,'3',d,u,'I','','Z',v,e,'','',q,d,'W','1',w,c,y,'',v,d,m,'V',y,'I','','R',o,'Z','S','',s,'Y','X',p,'5','I','','R',v,'Z',y,'4',=]=="VGhlIHF1aWNrIGJyb3duIGZveCBqdW1wcyBvdmVyIHRoZSBsYXp5IGRvZy4="
!       in file /Users/pmoura/logtalk/library/base64/tests.lgt between lines 56-61
!     base64url_parse_atom: failure (in 0.00087199999999976185 seconds)
!       test assertion failed: 'https//logtalk.org'=='https://logtalk.org'
!       in file /Users/pmoura/logtalk/library/base64/tests.lgt between lines 84-88
!     base64url_parse_chars: failure (in 0.0010760000000000769 seconds)
!       test assertion failed: [h,t,t,p,s,'',/,/,l,o,g,t,a,l,k,'.',o,r,g]=="https://logtalk.org"
!       in file /Users/pmoura/logtalk/library/base64/tests.lgt between lines 88-92
!     base64url_generate_chars: failure (in 0.00089800000000006541 seconds)
!       test assertion failed: [a,'','R','0',c,'','M','6','L',y,'9',s,b,'2',d,'0','Y','W',x,r,'L',m,'9',y,'Z',w]=="aHR0cHM6Ly9sb2d0YWxrLm9yZw"
!       in file /Users/pmoura/logtalk/library/base64/tests.lgt between lines 100-104
```

Notice how several chars are missing and replaced by the empty atom. This looks the same bug but may actually be two different bugs. Still digging...

**Comment by infradig** [comment]
Yeah, base64 I was accidentally  running lvm.

I noticed that character thing.

**Comment by pmoura** [comment]
In the case of the `base64` library, it's an `atom_codes/2` bug. Instrumenting the code to print the goal before and after, we get:

```logtalk
atom_codes(_70206,[86,71,104,108,73,72,70,49,97,87,78,114,73,71,74,121,98,51,100,117,73,71,90,118,101,67,66,113,100,87,49,119,99,121,66,118,100,109,86,121,73,72,82,111,90,83,66,115,89,88,112,53,73,71,82,118,90,121,52,61])
atom_codes('VhlIF1aWNrIJyb3duIZveqdW1wcyvdmVyIRoZSsYXp5IRvZy4=',[86,71,104,108,73,72,70,49,97,87,78,114,73,71,74,121,98,51,100,117,73,71,90,118,101,67,66,113,100,87,49,119,99,121,66,118,100,109,86,121,73,72,82,111,90,83,66,115,89,88,112,53,73,71,82,118,90,121,52,61])
```

But at the top-level:

```text
$ tpl
Trealla Prolog (c) Infradig 2020-2021, v1.10.11
?- atom_codes(A,[86,71,104,108,73,72,70,49,97,87,78,114,73,71,74,121,98,51,100,117,73,71,90,118,101,67,66,113,100,87,49,119,99,121,66,118,100,109,86,121,73,72,82,111,90,83,66,115,89,88,112,53,73,71,82,118,90,121,52,61]).
A = 'VGhlIHF1aWNrIGJyb3duIGZveCBqdW1wcyBvdmVyIHRoZSBsYXp5IGRvZy4='.
```

**Comment by infradig** [comment]
The second one direct from tpl is correct. So when running under Logtalk
it's does something different?

**Comment by pmoura** [comment]
Correct.

**Comment by pmoura** [comment]
On the `logtalk3/library/base64/base64.lgt` source file at line 83, I changed the clause to:

```logtalk
	generate(atom(Atom), Bytes) :-
		phrase(encode(Bytes), Codes),
		writeq(atom_codes(Atom, Codes)), nl,
		atom_codes(Atom, Codes),
		writeq(atom_codes(Atom, Codes)), nl,
		!.
```

**Comment by infradig** [comment]
All 'G' characters are missing!!!

**Comment by infradig** [comment]
It's nuts

V hlI F1aWNrI Jyb3duI Zve  qdW1wcy vdmVyIR oZS sYXp5I RvZy4=
VGhlIHF1aWNrIGJyb3duIGZveCBqdW1wcyBvdmVyIHRoZSBsYXp5IGRvZy4=
 G    H      G       G   CB       B       R   B      G

Why GHCBR?

**Comment by infradig** [comment]
OK!!!!! The missing letters are because the numbers are big integers
(unbounded)! So why is the question!

**Comment by infradig** [comment]
Ok, cracked it, it';s in fn_iso_shl_2 lines 1760+ trying to detect 64-bit
overflow into a bigint.

**Comment by pmoura** [comment]
Keeping the issue open as we still have the `lgt_write_term_3_25` test failure, caused by a different bug.

**Comment by pmoura** [comment]
Also, the latest fix seems to have introduced a regression in the `atom_codes/2` predicate:

```text
$ tpl
Trealla Prolog (c) Infradig 2020-2021, v1.10.13
?- atom_codes(abc, [a,b,c]).
false.
```

In this case, a `error(type_error(integer,a), ...)` exception is expected.

**Comment by pmoura** [comment]
Confirmed the `atom_codes/2`fix in 1.10.14.

**Comment by pmoura** [comment]
Confirmed the `write_term/2-3` bug fix in 1.10.63.
